### PR TITLE
feat(agw): Write per subscriber APN specific data into subscriberDB store when MME receives Update Location Answer

### PR DESCRIPTION
**Pull request by vroon2703**
## Summary

This PR has the following new functions: 

1. SQLite functions that allow adding subscribers 
2. A function to convert ula response to subscriber data which is then written into the sqlite database

## Test Plan
Unit tests for the function that converts ula to sub_data
Test that an attach/detach is still successful 

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/lib/s6a_proxy/S6aClient.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/lib/s6a_proxy/S6aClient.cpp]:
```suggestion
    // Only the first IP is assigned to the subscriber in the current
```

File: lte/gateway/c/core/oai/lib/store/sqlite.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/lib/store/sqlite.cpp]: Move this include to the top, as per the [Google Style guide](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)

File: lte/gateway/c/core/oai/lib/store/sqlite.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/lib/store/sqlite.cpp]:
```suggestion
    std::cout << "Subscriber DB location: " << db_location_list[shard] << std::endl;
```

File: lte/gateway/c/core/oai/lib/store/sqlite.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/lib/store/sqlite.cpp]: What's the purpose of this log? If it was only for debugging, please remove it, since it will be very hard to read.

File: lte/gateway/c/core/oai/lib/store/sqlite.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/lib/store/sqlite.cpp]: Remove debug log

File: lte/gateway/c/core/oai/lib/store/sqlite.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/lib/store/sqlite.cpp]: Remove debug log

Comment (ssanadhya) [reply lte/gateway/c/core/oai/lib/store/sqlite.cpp]: Instead, add a log stating that user APN info for user has been inserted.

File: lte/gateway/c/core/oai/lib/store/sqlite.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/lib/store/sqlite.cpp]: This function is not converting SubscriberData object to string, but the Subscriber ID of the given subscriber. Please rename it to `get_sid` or `get_imsi`.

File: lte/gateway/c/core/oai/test/lib/CMakeLists.txt
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/lib/CMakeLists.txt]: Remove extra newlines

**Review by ssanadhya — CHANGES_REQUESTED** [review]
Minor comments, please revise

File: lte/gateway/c/core/oai/lib/store/sqlite.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/lib/store/sqlite.cpp]:
```suggestion
    std::cout << "APN information for " << sid
```

File: lte/gateway/c/core/oai/lib/store/sqlite.hpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/lib/store/sqlite.hpp]:
```suggestion
  const char* _get_sid(const SubscriberData& subscriber_data);
```

File: lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go
Comment (ardzoht) [inline lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go]:
```suggestion
func GetCloudSubscriberDbEnabled(ctx context.Context) bool {
```

File: lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go
Comment (ardzoht) [inline lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go]: all these should be instead debug level?

File: lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go
Comment (ardzoht) [inline lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go]: I recommend to run `./build.py -c` under orc8r/cloud/docker to run `gofmt` on every Go file being updated here

**Review by ardzoht — CHANGES_REQUESTED** [review]

**Comment by ssanadhya** [comment]
@vroon2703 , switching the default return values to "false" for GetCloudSubscriberDbEnabled should fix the cloud tests.

**Review by vktng — CHANGES_REQUESTED** [review]
The Bazel build fails, because there is no `sqlite3.h` file on the virtual machines. After installing `libsqlite3-dev` this file will be present, but the build will still fail. To solve the issue you need to add the required flag to `bazel/external/system_libraries.BUILD`:

```
cc_library(
    name = "libsqlite3-dev",
    linkopts = ["-lsqlite3"],
)
```

And then include it in the `lte/gateway/c/core/BUILD.bazel` at the end of `MME_DEPS` like this:
`"@system_libraries//:libsqlite3-dev",`

**Comment by vroon2703** [comment]
> The Bazel build fails, because there is no `sqlite3.h` file on the virtual machines. After installing `libsqlite3-dev` this file will be present, but the build will still fail. To solve the issue you need to add the required flag to `bazel/external/system_libraries.BUILD`:
> 
> ```
> cc_library(
>     name = "libsqlite3-dev",
>     linkopts = ["-lsqlite3"],
> )
> ```
> 
> And then include it in the `lte/gateway/c/core/BUILD.bazel` at the end of `MME_DEPS` like this: `"@system_libraries//:libsqlite3-dev",`

Hey @vktng I made these changes to my PR (https://github.com/magma/magma/pull/13470/files#diff-5fef9f45d8fae0e6463b7bf908431d40936706fa860f0bb56002b3d00b77b285 and https://github.com/magma/magma/pull/13470/files#diff-5fe38d183eb821ecf919c6774f8b64769253b00d036d265dd853f95dcd7be484) but the BAZEL test still fails. Am I missing something else? Thanks!

**Comment by vktng** [comment]
> > The Bazel build fails, because there is no `sqlite3.h` file on the virtual machines. After installing `libsqlite3-dev` this file will be present, but the build will still fail. To solve the issue you need to add the required flag to `bazel/external/system_libraries.BUILD`:
> > ```
> > cc_library(
> >     name = "libsqlite3-dev",
> >     linkopts = ["-lsqlite3"],
> > )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And then include it in the `lte/gateway/c/core/BUILD.bazel` at the end of `MME_DEPS` like this: `"@system_libraries//:libsqlite3-dev",`
> 
> Hey @vktng I made these changes to my PR (https://github.com/magma/magma/pull/13470/files#diff-5fef9f45d8fae0e6463b7bf908431d40936706fa860f0bb56002b3d00b77b285 and https://github.com/magma/magma/pull/13470/files#diff-5fe38d183eb821ecf919c6774f8b64769253b00d036d265dd853f95dcd7be484) but the BAZEL test still fails. Am I missing something else? Thanks!

The dependency is still not on the VMs the tests are running on. You need to create a separate pull request where you add libsqlite3-dev as a dependency, merge it, and then this PR should pass.

**Review by ssanadhya — APPROVED** [review]
Lgtm, can be merged once CI issues are resolved.

**Comment by ssanadhya** [comment]
@vroon2703 , are you planning to move the changes to `lte/gateway/Vagrantfile` to a new PR. I see that change got removed from this PR.

> > > The Bazel build fails, because there is no `sqlite3.h` file on the virtual machines. After installing `libsqlite3-dev` this file will be present, but the build will still fail. To solve the issue you need to add the required flag to `bazel/external/system_libraries.BUILD`:
> > > ```
> > > cc_library(
> > >     name = "libsqlite3-dev",
> > >     linkopts = ["-lsqlite3"],
> > > )
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > And then include it in the `lte/gateway/c/core/BUILD.bazel` at the end of `MME_DEPS` like this: `"@system_libraries//:libsqlite3-dev",`
> > 
> > 
> > Hey @vktng I made these changes to my PR (https://github.com/magma/magma/pull/13470/files#diff-5fef9f45d8fae0e6463b7bf908431d40936706fa860f0bb56002b3d00b77b285 and https://github.com/magma/magma/pull/13470/files#diff-5fe38d183eb821ecf919c6774f8b64769253b00d036d265dd853f95dcd7be484) but the BAZEL test still fails. Am I missing something else? Thanks!
> 
> The dependency is still not on the VMs and the bazel-base container. You need to create a separate pull request where you add libsqlite3-dev as a dependency to them, merge it, and then this PR should pass. This should also resolve your issue with the failing mme test job. If you have more Bazel related questions you can find us in the [#bazel](https://magmacore.slack.com/archives/C033S1CEAUU) Slack channel.

**Comment by ssanadhya** [comment]
@rdefosse, could you please help fix Magma-OAI-Jenkins CI failure here? I see from the [logs](https://jenkins-oai.eurecom.fr/job/MAGMA-MME-production/17446/artifact/build_results_magma_oai_mme.html) that it is missing the libsqlite3-dev package on the container running in CI. Which docker files need to be updated for this CI job?

**Comment by rdefosse** [comment]
@ssanadhya @vroon2703 

Sorry I did not see the notification.

Working on it.

**Comment by rdefosse** [comment]
@vroon2703 

There are 4 files you SHALL modify for this PR to pass on `OAI` pipeline.

I've upgraded the base images to include `libsqlite3x` and `libsqlite3x-dev`

```git
diff --git a/ci-scripts/docker/Dockerfile.mme.ci.rhel8 b/ci-scripts/docker/Dockerfile.mme.ci.rhel8
index aee5fbbaad..7ab57caba8 100644
--- a/ci-scripts/docker/Dockerfile.mme.ci.rhel8
+++ b/ci-scripts/docker/Dockerfile.mme.ci.rhel8
@@ -65,6 +65,7 @@ RUN yum update -y && \
       libubsan \
       libasan \
       liblsan \
+      libsqlite3x \
       psmisc \
       procps-ng \
       tcpdump \
diff --git a/ci-scripts/docker/Dockerfile.mme.ci.ubuntu18 b/ci-scripts/docker/Dockerfile.mme.ci.ubuntu18
index 188519853d..21f7fe5935 100644
--- a/ci-scripts/docker/Dockerfile.mme.ci.ubuntu18
+++ b/ci-scripts/docker/Dockerfile.mme.ci.ubuntu18
@@ -70,6 +70,7 @@ RUN apt-get update && \
       net-tools \
       tshark \
       tzdata \
+      libsqlite3-0 \
   && rm -rf /var/lib/apt/lists/*
 
 # Copy pre-built shared object files
diff --git a/lte/gateway/docker/mme/Dockerfile.rhel8 b/lte/gateway/docker/mme/Dockerfile.rhel8
index 2d9f2f805b..88b96b560e 100644
--- a/lte/gateway/docker/mme/Dockerfile.rhel8
+++ b/lte/gateway/docker/mme/Dockerfile.rhel8
@@ -45,6 +45,8 @@ RUN rm /etc/rhsm-host && \
     protobuf-compiler \
     ruby \
     ruby-devel \
+    libsqlite3x \
+    libsqlite3x-devel \
   && ln -s /usr/bin/python3 /usr/bin/python \
   && ln -s /usr/bin/cmake3 /usr/bin/cmake
 
@@ -427,6 +429,7 @@ RUN yum update -y && \
       libubsan \
       libasan \
       liblsan \
+      libsqlite3x \
       psmisc \
       procps-ng \
       tcpdump \
diff --git a/lte/gateway/docker/mme/Dockerfile.ubuntu18.04 b/lte/gateway/docker/mme/Dockerfile.ubuntu18.04
index 23581cb0d9..596c2b2689 100644
--- a/lte/gateway/docker/mme/Dockerfile.ubuntu18.04
+++ b/lte/gateway/docker/mme/Dockerfile.ubuntu18.04
@@ -17,7 +17,7 @@ RUN [ "/bin/bash", "-c", "echo \"Install general purpose packages\" && \
     apt-get update && \
     DEBIAN_FRONTEND=noninteractive apt-get upgrade -y && \
     DEBIAN_FRONTEND=noninteractive apt-get install -y gnupg wget software-properties-common autoconf automake \
-    libtool curl make g++ unzip git build-essential autoconf libtool pkg-config \
+    libtool curl make g++ unzip git build-essential autoconf libtool pkg-config libsqlite3-dev libsqlite3-0 \
     gcc-7 g++-7 apt-transport-https ca-certificates apt-utils vim redis-server tzdata \
     libssl-dev ninja-build golang python2.7 automake perl libgmp3-dev clang-format-7 && \
     echo \"Configure C/C++ compiler v7.5 as primary\" && \
@@ -275,6 +275,7 @@ RUN apt-get update && \
       net-tools \
       tshark \
       tzdata \
+      libsqlite3-0 \
   && rm -rf /var/lib/apt/lists/*
 
 # Copy pre-built shared object files
```

**Comment by ssanadhya** [comment]
Thanks for the detailed response @rdefosse ! I have added these changes as a separate PR https://github.com/magma/magma/pull/13575 . Once that is merged, this PR should be green on CI. 
cc: @vroon2703 


> @vroon2703
> 
> There are 4 files you SHALL modify for this PR to pass on `OAI` pipeline.
> 
> I've upgraded the base images to include `libsqlite3x` and `libsqlite3x-dev`
> 
> ```
> diff --git a/ci-scripts/docker/Dockerfile.mme.ci.rhel8 b/ci-scripts/docker/Dockerfile.mme.ci.rhel8
> index aee5fbbaad..7ab57caba8 100644
> --- a/ci-scripts/docker/Dockerfile.mme.ci.rhel8
> +++ b/ci-scripts/docker/Dockerfile.mme.ci.rhel8
> @@ -65,6 +65,7 @@ RUN yum update -y && \
>        libubsan \
>        libasan \
>        liblsan \
> +      libsqlite3x \
>        psmisc \
>        procps-ng \
>        tcpdump \
> diff --git a/ci-scripts/docker/Dockerfile.mme.ci.ubuntu18 b/ci-scripts/docker/Dockerfile.mme.ci.ubuntu18
> index 188519853d..21f7fe5935 100644
> --- a/ci-scripts/docker/Dockerfile.mme.ci.ubuntu18
> +++ b/ci-scripts/docker/Dockerfile.mme.ci.ubuntu18
> @@ -70,6 +70,7 @@ RUN apt-get update && \
>        net-tools \
>        tshark \
>        tzdata \
> +      libsqlite3-0 \
>    && rm -rf /var/lib/apt/lists/*
>  
>  # Copy pre-built shared object files
> diff --git a/lte/gateway/docker/mme/Dockerfile.rhel8 b/lte/gateway/docker/mme/Dockerfile.rhel8
> index 2d9f2f805b..88b96b560e 100644
> --- a/lte/gateway/docker/mme/Dockerfile.rhel8
> +++ b/lte/gateway/docker/mme/Dockerfile.rhel8
> @@ -45,6 +45,8 @@ RUN rm /etc/rhsm-host && \
>      protobuf-compiler \
>      ruby \
>      ruby-devel \
> +    libsqlite3x \
> +    libsqlite3x-devel \
>    && ln -s /usr/bin/python3 /usr/bin/python \
>    && ln -s /usr/bin/cmake3 /usr/bin/cmake
>  
> @@ -427,6 +429,7 @@ RUN yum update -y && \
>        libubsan \
>        libasan \
>        liblsan \
> +      libsqlite3x \
>        psmisc \
>        procps-ng \
>        tcpdump \
> diff --git a/lte/gateway/docker/mme/Dockerfile.ubuntu18.04 b/lte/gateway/docker/mme/Dockerfile.ubuntu18.04
> index 23581cb0d9..596c2b2689 100644
> --- a/lte/gateway/docker/mme/Dockerfile.ubuntu18.04
> +++ b/lte/gateway/docker/mme/Dockerfile.ubuntu18.04
> @@ -17,7 +17,7 @@ RUN [ "/bin/bash", "-c", "echo \"Install general purpose packages\" && \
>      apt-get update && \
>      DEBIAN_FRONTEND=noninteractive apt-get upgrade -y && \
>      DEBIAN_FRONTEND=noninteractive apt-get install -y gnupg wget software-properties-common autoconf automake \
> -    libtool curl make g++ unzip git build-essential autoconf libtool pkg-config \
> +    libtool curl make g++ unzip git build-essential autoconf libtool pkg-config libsqlite3-dev libsqlite3-0 \
>      gcc-7 g++-7 apt-transport-https ca-certificates apt-utils vim redis-server tzdata \
>      libssl-dev ninja-build golang python2.7 automake perl libgmp3-dev clang-format-7 && \
>      echo \"Configure C/C++ compiler v7.5 as primary\" && \
> @@ -275,6 +275,7 @@ RUN apt-get update && \
>        net-tools \
>        tshark \
>        tzdata \
> +      libsqlite3-0 \
>    && rm -rf /var/lib/apt/lists/*
>  
>  # Copy pre-built shared object files
> ```

**Review by six66vivi — APPROVED** [review]

**Comment by ssanadhya** [comment]
@vroon2703 , Changes for Magma-OAI-Jenkins are merged in https://github.com/magma/magma/pull/13575. Could you please rebase this PR on the latest master to re-trigger CI checks?

**Review by vktng — COMMENTED** [review]
Bazel changes LGTM, waiting for the rebase before approving.

**Review by ssanadhya — APPROVED** [review]
Lgtm

**Comment by ssanadhya** [comment]
@ardzoht , @vktng , could you please review and confirm if the requested changes have been addressed on this PR?

**Review by vktng — APPROVED** [review]
Bazel changes LGTM.

**Review by ardzoht — APPROVED** [review]
lgtm

File: orc8r/gateway/configs/service_registry.yml
Comment (ssanadhya) [inline orc8r/gateway/configs/service_registry.yml]: @vroon2703 , is this a valid change?

**Review by ssanadhya — CHANGES_REQUESTED** [review]

**Review by LKreutzer — APPROVED** [review]
Bazel changes reviewed in pairing. I do not see any changes right now that require CI codeowner approval

**Review by ssanadhya — APPROVED** [review]

File: lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go
Comment (emakeev) [inline lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go:124]: It looks like that if cloud DB is enabled, we should just return nil, nil. Why not to simplify & just return (nil, nil) on LN 79 above?

File: lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go
Comment (emakeev) [inline lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go:277]: Move it up to the beginning of the function. There is no reason to waste resources on building the changeset if it'll get discarded anyway.

File: lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go
Comment (emakeev) [inline lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go:306]: same, move it up to the beginning, no reason to compile the page if it's not needed

File: lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go
Comment (emakeev) [inline lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go:114]: GetCloudSubscriberDbEnabled(ctx) is a bit heavy, it has to lookup GW info, load network configs (all DB lookups/reads). I think it'd make sense to pass cloudflag bool to getSubscribersChangeset as a parameter instead of making another redundant call to GetCloudSubscriberDbEnabled() inside getSubscribersChangeset().

File: lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go
Comment (emakeev) [inline lte/cloud/go/services/subscriberdb/servicers/southbound/subscriberdb_servicer.go:259]: change ctx to cloudSubscriberDBEnabled bool.

**Review by emakeev — APPROVED** [review]
Please consider prior suggestions from inline comments before merging.

**Comment by ssanadhya** [comment]
@vroon2703 , you will need to rebase onto latest master to re-trigger the `check-semantic-pr` check.

**Comment by ssanadhya** [comment]
Force merging this PR since it was blocked on semantic PR check for several days. The PR title looks semantically correct.

**Pull request merged**
